File: backend/internal/contacts/contacts.go
// SPDX-FileCopyrightText: 2026 Vladislav Janeček
// SPDX-License-Identifier: AGPL-3.0-only

// Package contacts is the address-book side of recipient completion: the
// types the core consumes and the Directory interface a system
// address-book client (internal/contacts/eds) implements. It depends on
// nothing of its own, so core and the clients can share it without a cycle.
package contacts

import "context"

// Book is one address book a directory offers for completion.
type Book struct {
	UID     string // the source's uid; a single D-Bus path segment
	Name    string // display name, untrusted text
	Backend string // "microsoft365", "carddav", "google", "local", …
	// Collection is the parent source's uid, "" for a stand-alone book.
	// GOAAccountID and Email are the collection's identity when it has one
	// (a GNOME Online Accounts account id, and its normalised address).
	Collection   string
	GOAAccountID string
	Email        string
}

// Contact is one card's address. A card with several addresses yields one
// Contact per address.
type Contact struct {
	Name    string // untrusted display text; "" when the card has none
	Address string // normalised, syntactically valid
	Book    string // display name of the book it came from, untrusted
}

// Directory is what core needs from a system address-book client. Books
// lists the address books that take part in completion; Search queries
// the given books. Both report api.CodeUnavailable when there is no
// session bus or no address-book service, which core treats as "no
// address-book results", never as a failed search.
type Directory interface {
	Books(ctx context.Context) ([]Book, error)
	Search(ctx context.Context, books []Book, query string, limit int) ([]Contact, error)
}

File: macos/Sources/MalachiCore/API/Contacts.swift
// SPDX-FileCopyrightText: 2026 Vladislav Janeček
// SPDX-License-Identifier: GPL-3.0-or-later

// Recipient completion (docs/api.md §4.11; types.go Contact).

import Foundation

/// api.Contact: one recipient suggestion. `name` and `book` are untrusted
/// display text.
public struct Contact: Codable, Sendable, Equatable {
    public var name: String?
    /// Normalised, syntactically valid.
    public var address: String
    public var source: ContactSource
    /// The address book the contact came from; nil for a collected address.
    public var book: String?

    public init(name: String? = nil, address: String, source: ContactSource, book: String? = nil) {
        self.name = name
        self.address = address
        self.source = source
        self.book = book
    }
}

/// api.ContactSearchParams. `limit` nil = `API.Limits.defaultContactLimit`,
/// clamped to `maxContactLimit`; `query` at most `maxContactQueryBytes`.
public struct ContactSearchParams: Codable, Sendable, Equatable {
    public var accountId: AccountID
    public var query: String
    public var limit: Int?

    public init(accountId: AccountID, query: String, limit: Int? = nil) {
        self.accountId = accountId
        self.query = query
        self.limit = limit
    }
}

/// api.ContactSearchResult: ranked best first.
public struct ContactSearchResult: Codable, Sendable, Equatable {
    @NullAsEmpty public var contacts: [Contact]

    public init(contacts: [Contact]) {
        self.contacts = contacts
    }
}
